pkg/controller/ec2/routetable/controller.go

```diff
@@ -237,33 +250,110 @@ func (e *external) Delete(ctx context.Context, mgd resource.Managed) error {
+			}).Send(ctx)
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (e *external) createRoutes(ctx context.Context, tableID string, desired []v1alpha5.Route, observed []v1alpha5.RouteState) error { // nolint:gocyclo
 	for _, rt := range desired {
 		isObserved := false
 		for _, ob := range observed {
-			if ob.GatewayID == aws.StringValue(rt.GatewayID) && ob.DestinationCIDRBlock == aws.StringValue(rt.DestinationCIDRBlock) {
+			if ob.DestinationCIDRBlock == aws.StringValue(rt.DestinationCIDRBlock) && (ob.GatewayID == aws.StringValue(rt.GatewayID) ||
+				ob.InstanceID == aws.StringValue(rt.InstanceID) ||
+				ob.LocalGatewayID == aws.StringValue(rt.LocalGatewayID) ||
+				ob.NatGatewayID == aws.StringValue(rt.NatGatewayID) ||
+				ob.NetworkInterfaceID == aws.StringValue(rt.NetworkInterfaceID) ||
+				ob.TransitGatewayID == aws.StringValue(rt.TransitGatewayID) ||
+				ob.VpcPeeringConnectionID == aws.StringValue(rt.VpcPeeringConnectionID)) {
 				isObserved = true
 				break
 			}
 		}
 		// if the route is already created, skip it
 		if !isObserved {
 			_, err := e.client.CreateRouteRequest(&awsec2.CreateRouteInput{
-				RouteTableId:         aws.String(tableID),
-				DestinationCidrBlock: rt.DestinationCIDRBlock,
-				GatewayId:            rt.GatewayID,
+				RouteTableId:             aws.String(tableID),
+				DestinationCidrBlock:     rt.DestinationCIDRBlock,
+				DestinationIpv6CidrBlock: rt.DestinationIPV6CIDRBlock,
+				GatewayId:                rt.GatewayID,
+				InstanceId:               rt.InstanceID,
+				LocalGatewayId:           rt.LocalGatewayID,
+				NatGatewayId:             rt.NatGatewayID,
+				NetworkInterfaceId:       rt.NetworkInterfaceID,
+				TransitGatewayId:         rt.TransitGatewayID,
+				VpcPeeringConnectionId:   rt.VpcPeeringConnectionID,
 			}).Send(ctx)
 
 			if err != nil {
-				return awsclient.Wrap(err, errCreateRoute)
+				return err
 			}
 		}
 	}
+	return nil
+}
+
+func (e *external) reconcileRoutes(ctx context.Context, tableID string, desired []v1alpha5.Route, observed []v1alpha5.RouteState) error {
+
+	if err := e.deleteRoutes(ctx, tableID, desired, observed); err != nil {
+		return awsclient.Wrap(err, errDeleteRoute)
+	}
+	if err := e.createRoutes(ctx, tableID, desired, observed); err != nil {
+		return awsclient.Wrap(err, errCreateRoute)
+	}
 
 	return nil
 }
 
-func (e *external) createAssociations(ctx context.Context, tableID string, desired []v1alpha4.Association, observed []v1alpha4.AssociationState) error {
+func (e *external) removeAssociations(ctx context.Context, desired []v1alpha5.Association, observed []v1alpha5.AssociationState) error {
+	for _, asc := range observed {
+		found := false
+		for _, ds := range desired {
+			if asc.SubnetID == aws.StringValue(ds.SubnetID) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No longer needed so delete
+			_, err := e.client.DisassociateRouteTableRequest(&awsec2.DisassociateRouteTableInput{
```

Comment:
Can we reuse `deleteAssociations` function here?

Reply:
This seems to be marked as resolved. Is there a specific reason why we can't accumulate all associations into a `[]v1alpha4.AssociationState` and pass that to `deleteAssociations` after the loop finishes? Having one implementation is better because for example here `NotFound` case is not ignored while `deleteAssociations` takes care of that.

Reply:
Oops sorry swear I had done this already. Went ahead and got it added.